### PR TITLE
Oasis app version 2.x support

### PR DESCRIPTION
**Pull request by pro-wh**
enable app version 2.x. in this PR, getting the public key and signing

also some error handling fixes

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -61,6 +61,7 @@ export default class OasisApp {
     switch (this.versionResponse.major) {
       case 0:
       case 1:
+      case 2:
```
Comment (pro-wh) [inline src/index.js:64]: is this correct? that version 2 uses the same serialize path implementation?

File: src/index.js [hunk 2]
```diff
@@ -103,6 +104,7 @@ export default class OasisApp {
 
   async signGetChunks(path, context, message) {
     const serializedPath = await this.serializePath(path);
+    if ("return_code" in serializedPath && serializedPath.return_code !== 0x9000) return serializedPath;
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply src/index.js:107]: yeah, let's add these comments

File: src/index.js [hunk 3]
```diff
@@ -242,8 +248,10 @@ export default class OasisApp {
 
   async sign(path, context, message) {
     const chunks = await this.signGetChunks(path, context, message);
+    if ("return_code" in chunks && chunks.return_code !== 0x9000) return chunks;
 
-    return this.signSendChunk(1, chunks.length, chunks[0], [0x9000]).then(async (response) => {
+    return this.signSendChunk(1, chunks.length, chunks[0]).then(async (response) => {
+      if (response.return_code !== 0x9000) return response;
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply src/index.js:254]: I had found examples similar to this one-line form, but let's go with the three line form

**Comment by pro-wh** [comment]
> Looking at source code in files in `src/` in this repo and source code in files in [`js/src/` in Ledger Oasis app repo](https://github.com/Zondax/ledger-oasis/tree/v2.3.1/js/src), there is a lot of copying/duplication.
> 
> What would be the best way to avoid that?

I didn't know about there being a copy of this module in the repo of the app itself. Is this intended? How did the two copies com to be?

> BTW, what is the "read public key" part of the PRs title?

This PR had originally addressed the read public key functionality. But I had since found that the signing also needed updating, and I later added that here too. I should just remove this part of the title

**Comment by tjanez** [comment]
> > Looking at source code in files in `src/` in this repo and source code in files in [`js/src/` in Ledger Oasis app repo](https://github.com/Zondax/ledger-oasis/tree/v2.3.1/js/src), there is a lot of copying/duplication.
> > What would be the best way to avoid that?
> 
> I didn't know about there being a copy of this module in the repo of the app itself. Is this intended? How did the two copies com to be?

I also don't know the genesis of this :slightly_smiling_face:.

Can you open an issue in this repo to track this unusual code duplication and we can tackle that separately from this PR?

> > BTW, what is the "read public key" part of the PRs title?
> 
> This PR had originally addressed the read public key functionality. But I had since found that the signing also needed updating, and I later added that here too. I should just remove this part of the title

Aha, I see. Yeah, please, just update the title to avoid confusion :slightly_smiling_face:.

**Comment by pro-wh** [comment]
title changed, filed #288

gonna make some changes to address review comments next

File: src/index.js [hunk 4]
```diff
@@ -231,6 +236,7 @@ export default class OasisApp {
     switch (this.versionResponse.major) {
       case 0:
       case 1:
+      case 2:
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply src/index.js:239]: nice

File: src/index.js (hunk 1 above)
Comment (pro-wh) [reply src/index.js:64]: wow github breaks that link by removing the `#`

`https://github.com/Zondax/ledger-oasis/compare/v1.8.2...v2.3.1#diff-1fe0c5c30620ad391d8a90a7580d5995a738548393e38f7329c90429691a0ef6`

reposting in code block

Comment (pro-wh) [reply src/index.js:64]: oh the 5-component path uses unhardened derivations for the last one?

**Comment by pro-wh** [comment]
weird, github sets co-authored-by line to the user who started the thread, not the author of the suggestion. gonna fix that manually

File: src/index.js (hunk 1 above)
Comment (pro-wh) [reply src/index.js:64]: anyway, verified that the new app is compatible. thanks for the reference

File: src/common.js [hunk 5]
```diff
@@ -52,6 +52,8 @@ export function errorCodeToString(statusCode) {
 }
 
 export function processErrorResponse(response) {
+  // Leave non-Ledger errors as they are.
+  if (!("statusCode" in response)) throw response;
```
[the remark this answers is not in the document]
Comment (pro-wh) [reply src/common.js]: let's leave this as is, because there's good locality between the reference here and there in the nearly following lines

**Review by tjanez — APPROVED** [review]
Looks good! Thanks, @pro-wh!

**Comment by pro-wh** [comment]
thanks for the help!

**Pull request merged**